File: test/test.EMGLLF.c
#include "EMGLLF.h"
#include "test_utils.h"
#include <stdlib.h>

int main(int argc, char** argv)
{
	int* dimensions = readArray_int("dimensions");
	int n = dimensions[0];
	int p = dimensions[1];
	int m = dimensions[2];
	int k = dimensions[3];
	free(dimensions);

	////////////
	// INPUTS //
	Real* phiInit = readArray_real("phiInit");
	Real* rhoInit = readArray_real("rhoInit");
	Real* piInit = readArray_real("piInit");
	Real* gamInit = readArray_real("gamInit");
	int mini = read_int("mini");
	int maxi = read_int("maxi");
	Real gamma = read_real("gamma");
	Real lambda = read_real("lambda");
	Real* X = readArray_real("X");
	Real* Y = readArray_real("Y");
	Real tau = read_real("tau");
	////////////

	/////////////
	// OUTPUTS //
	Real* phi = (Real*)malloc(p*m*k*sizeof(Real));
	Real* rho = (Real*)malloc(m*m*k*sizeof(Real));
	Real* pi = (Real*)malloc(k*sizeof(Real));
	Real llh;
	Real* S = (Real*)malloc(p*m*k*sizeof(Real));
	int* affec = (int*)malloc(n*sizeof(int));
	/////////////

	////////////////////
	// Call to EMGLLF //
	EMGLLF_core(phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,lambda,X,Y,tau,
		phi,rho,pi,&llh,S,affec,
		n,p,m,k);
	////////////////////

	free(phiInit);
	free(rhoInit);
	free(piInit);
	free(gamInit);
	free(X);
	free(Y);

	// Compare to reference outputs
	Real* ref_phi = readArray_real("phi");
	compareArray_real("phi", phi, ref_phi, p*m*k);
	free(phi);
	free(ref_phi);

	Real* ref_rho = readArray_real("rho");
	compareArray_real("rho", rho, ref_rho, m*m*k);
	free(rho);
	free(ref_rho);

	Real* ref_pi = readArray_real("pi");
	compareArray_real("pi", pi, ref_pi, k);
	free(pi);
	free(ref_pi);

	Real ref_llh = read_real("llh");
	compareArray_real("llh", &llh, &ref_llh, 1);

	Real* ref_S = readArray_real("S");
	compareArray_real("S", S, ref_S, p*m*k);
	free(S);
	free(ref_S);

	int* ref_affec = readArray_int("affec");
	compareArray_int("affec", affec, ref_affec, n);
	free(affec);
	free(ref_affec);

	return 0;
}

File: test/test_EMGLLF.R
library(valse)
testFolder = "data/"

# NOTE: R typing is really terrible. as.double as.matrix ...and so on; don't remove.

#get inputs
npmk = as.matrix(read.table(paste(testFolder,"dimensions",sep="")))
n = npmk[1]; p=npmk[2]; m=npmk[3]; k=npmk[4]
phiInit = array(as.double(as.matrix(read.table(paste(testFolder,"phiInit",sep="")))), dim=c(p,m,k))
rhoInit = array(as.double(as.matrix(read.table(paste(testFolder,"rhoInit",sep="")))), dim=c(m,m,k))
piInit = as.double(as.matrix(read.table(paste(testFolder,"piInit",sep="")))[,])
gamInit = matrix(as.double(as.matrix(read.table(paste(testFolder,"gamInit",sep="")))), n,k)
mini = as.integer(as.matrix(read.table(paste(testFolder,"mini",sep="")))[1])
maxi = as.integer(as.matrix(read.table(paste(testFolder,"maxi",sep="")))[1])
gamma = as.double(as.matrix(read.table(paste(testFolder,"gamma",sep="")))[1])
lambda = as.double(as.matrix(read.table(paste(testFolder,"lambda",sep="")))[1])
X = matrix(as.double(as.matrix(read.table(paste(testFolder,"X",sep="")))), n,p)
Y = matrix(as.double(as.matrix(read.table(paste(testFolder,"Y",sep="")))), n,m)
eps = as.double(as.matrix(read.table(paste(testFolder,"eps",sep="")))[1])

#get outputs
phi = array(as.double(as.matrix(read.table(paste(testFolder,"phi",sep="")))), dim=c(p,m,k))
rho = array(as.double(as.matrix(read.table(paste(testFolder,"rho",sep="")))), dim=c(m,m,k))
pi = as.double(as.matrix(read.table(paste(testFolder,"pi",sep="")))[,])
llh = as.double(as.matrix(read.table(paste(testFolder,"llh",sep="")))[1])
S = array(as.double(as.matrix(read.table(paste(testFolder,"S",sep="")))), dim=c(p,m,k))
affec = as.double(as.matrix(read.table(paste(testFolder,"affec",sep="")))[,])

res = valse::EMGLLF(
	phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,lambda,X,Y,eps,fast=TRUE)

#compare outputs
nd=7 #number of digits
print( all(round(phi,nd) == round(res$phi,nd)) )
print( all(round(rho,nd) == round(res$rho,nd)) )
print( all(round(pi,nd) == round(res$pi,nd)) )
print( all(round(llh,nd) == round(res$llh,nd)) )
print( all(round(S,nd) == round(res$S,nd)) )
print( all(affec == res$affec) )
